a3p-integration/proposals/z:acceptance/psm.test.js

```diff
@@ -0,0 +1,274 @@
+  getPsmMetrics,
+  implementPsmGovParamChange,
+  initializeNewUser,
+  maxMintBelowLimit,
+} from './test-lib/psm-lib.js';
+import { getBalances } from './test-lib/utils.js';
+
+// Export these from synthetic-chain?
+export const USDC_DENOM = process.env.USDC_DENOM
+  ? process.env.USDC_DENOM
+  : 'no-denom';
+export const PSM_PAIR = process.env.PSM_PAIR?.replace('.', '-');
+
+const psmTestSpecs = {
+  govParams: {
+    giveMintedFeeVal: 10n, // in %
+    wantMintedFeeVal: 10n, // in %
+    mintLimit: 500n * 1_000_000n, // in IST
+    deadline: 1, // in minutes
+  },
+  psmInstance: `psm-${PSM_PAIR}`,
+  // @ts-expect-error we assume PSM_PAIR is set because of synthetic-chain environment
+  anchor: PSM_PAIR.split('-')[1],
+  newUser: {
+    name: 'new-psm-trader',
+    fund: {
+      denom: USDC_DENOM,
+      value: '300000000', // 300 USDC_axl
+    },
+  },
+  otherUser: {
+    name: 'gov1',
+    fund: {
+      denom: USDC_DENOM,
+      value: '1000000000', // 1000 USDC_axl
+    },
+    toIst: {
+      value: 500, // in IST
+    },
+  },
+  toIst: {
+    value: 50, // in IST
+  },
+  fromIst: {
+    value: 50, // in USDC_axl
+  },
+};
+
+test.serial('change gov params', async t => {
+  await implementPsmGovParamChange(
+    {
+      address: GOV1ADDR,
+      instanceName: psmTestSpecs.psmInstance,
+      newParams: psmTestSpecs.govParams,
+      deadline: psmTestSpecs.govParams.deadline,
+    },
+    { committeeAddrs: [GOV1ADDR, GOV2ADDR, GOV3ADDR], position: 0 },
+  );
+
+  // Replace when https://github.com/Agoric/agoric-sdk/pull/10171 is in
+  await waitForBlock(3);
+  await checkGovParams(
+    t,
+    {
+      GiveMintedFee: {
+        type: 'ratio',
+        value: {
+          numerator: { value: psmTestSpecs.govParams.giveMintedFeeVal * 100n }, // convert to bps
+        },
+      },
+      WantMintedFee: {
+        type: 'ratio',
+        value: {
+          numerator: { value: psmTestSpecs.govParams.wantMintedFeeVal * 100n }, // convert to bps
+        },
+      },
+      MintLimit: {
+        type: 'amount',
+        value: {
+          value: psmTestSpecs.govParams.mintLimit,
+        },
+      },
+    },
+    psmTestSpecs.psmInstance.split('-')[2],
+  );
+});
+
+test.serial('initialize new user', async t => {
+  const {
+    newUser: { name, fund },
+  } = psmTestSpecs;
+
+  await initializeNewUser(name, fund);
+  // Replace when https://github.com/Agoric/agoric-sdk/pull/10171 is in
+  await waitForBlock(3);
+
+  await checkUserInitializedSuccessfully(name, fund);
+  t.pass();
+});
+
+test.serial('swap into IST', async t => {
+  const {
+    newUser: { name },
+    anchor,
+    toIst,
+    govParams: { wantMintedFeeVal },
+  } = psmTestSpecs;
+
+  const psmTrader = await getUser(name);
+
+  const [metricsBefore, balances] = await Promise.all([
+    getPsmMetrics(anchor),
+    getBalances([psmTrader]),
+  ]);
+
+  const balancesBefore = await adjustBalancesIfNotProvisioned(
+    balances,
+    psmTrader,
+  );
```

Comment:
I find this `adjustBalancesIfNotProvisioned` logic hard to follow. In the "not provisioned" case, `balancesBefore` isn't actually balances before.
 
perhaps...

```suggestion
  const alreadyProvisioned  = (await allWalletAddresses()).includes(psmTrader);
  const balancesBefore = await getBalances(psmTrader);
```

and then pass `alreadyProvisioned` to `checkSwapSucceeded`?

Reply:
ignored for now